2015/9-easy-exceptions-and-logging.pod

```diff
@@ -0,0 +1,293 @@
+
+Messages are automatically saved in the session, in a key called C<messages>.
+By default, not all messages are saved, as you probably don't want the user
+seeing your debug messages. The default configuration of which messages are
+sent should work for most people, but can be configured if required.
+
+To display the messages, you can do something like this in your layout template:
+
+ [% FOREACH msg IN session.messages %]
+   <p>[% msg.toString | html_entity %]</p>
+ [% END %]
+
+Each message is an object, and can be stringified as above.
+
+The object includes a function to automatically select a sensible Bootstrap
+color, so if you're using Bootstrap in your application, you could do this:
+
+ [% FOREACH msg IN session.messages %]
+   <div class="alert alert-[% msg.bootstrap_color %]">
+     [% msg.toString | html_entity %]
+   </div>
+ [% END %]
+
+=item * Using the message keywords without using C<Dancer2>
+
+In any other module, you can now include C<Log::Report> and use the same
+keywords/functions to send messages to your application (with the exception
+of C<success> - although there is a workaround for that if needed).
+
+=item * Clean exception handling
+
+The module will also, by default, handle any application exceptions cleanly.
+Instead of sending a 500 error, it will forward to an appropriate page, with
+the message rendered in the template as above. More on this to follow later.
+
+If the application is running with C<show_errors> enabled, then the full
+error will be sent in the message. If C<show_errors> is disabled (such as
+in a production environment), a generic error message is shown (the full
+error will have been sent to the console, or whichever backend is configured).
+
+=back
+
+So in summary, by including a couple of modules, we can easily send messages to
+a template, and we are handling application exceptions in a user-friendly manner,
+whilst still capturing the full error information.
+
+=head2 Handling user errors easily and cleanly
+
+Handling user errors can be a bit boring and tedious. We'll talk about a way
+to do so that requires very little code.
+
+First, let's have a look at how we might normally do it. I won't go into
+this in any detail, but you will probably recognise the concept. Maybe
+something like this:
+
+ package MyApp;
+
+ post '/settings' => sub {
+   my $message;
+   try {
+     MyApp::Settings->update($settings);
+     $message = "Settings updated successfully";
+   }
+   catch {
+     $message = $@;
+   };
+ };
+ 
+ package MyApp::Settings;
+
+ sub update {
+   ...
+   die "Invalid email address" unless ...
+ }
+
+That works fine to a degree, but the code can easily get more verbose than that,
+and even that alone can get a little tedious to keep typing.
+
+How would we do that with L<Dancer2::Plugin::LogReport>?
+
+ package MyApp;
+
+ post '/settings' => sub {
+   if (process sub{ MyApp::Settings::update($settings) }) {
+     success "Settings updated successfully";
+   }
+ };
```

Comment:
Actually, for this example, that could be it (hence the beauty of the plugin - no code required to handle the error message, it will already have been done). I've added that as a comment to the code.